### ISSUE TITLE
For CUT&RUN data

### ISSUE THREAD

**Issue by dpwlek1997** [loosolab/TF-COMB#68]
Hi, I am currently working on Cut&RUN data (with custom reference) with two different histone antibodies and trying to apply them to TFcomb. Here is some start codes as documentation. I used H3K4me3 antibody binding sites or called peaks for a bed file with Gene name annotation from Homer. But it resulted in zero rules. Can someone guide me through it? 

Thanks! 

```python
# set up an empty CombOb
C = CombObj()

# read TFBS from bed files
C.TFBS_from_bed("../data/A1_H3K4me3.bed") 
C.TFBS[:10]

# perform market basket analysis = co-occurrence analysis of the sites.
C.count_within()
C.market_basket()
```

**Comment by vheger** [comment]
Hi @dpwlek1997,
 
thank you for your interest in TF-COMB.  

For your Questions regarding the Input data: TF-COMB can be used in two ways 
1.) Either you provide the precalculated binding sites as a bed file (like in our example file [TFBS from chip](https://raw.githubusercontent.com/loosolab/TF-COMB/main/data/GM12878_hg38_chr4_TF_chipseq.bed)). 
The required format here is `1:chr 2:start 3:end 4:name 5..n:optional columns`. Please note the **name** in the fourth column.
 
2.) Or you provide peaks/regions and a list of motifs TF-COMB should scan for (like in our example files [peaks](https://github.com/loosolab/TF-COMB/blob/main/data/GM12878_hg38_chr4_ATAC_peaks.bed) and [motifs](https://github.com/loosolab/TF-COMB/blob/main/data/HOCOMOCOv11_HUMAN_motifs.txt)). Required format for bed file is `1:chr 2:start 3:end 4..n:optional columns`. Motifs can be given as JASPAR/MEME-style motifs or a tobias-format.
Please note: for the second use case, you need to use a different function as described in this [example](https://github.com/loosolab/TF-COMB/blob/main/examples/TFBS_from_motifs.ipynb).

Could you please check if the format of your .bed file is correct (e.g. no additional columns upfront like peakID from homer)  and you use the right use-case ([TFBS_from_bed()](https://tf-comb.readthedocs.io/en/latest/API/tfcomb.objects.html#tfcomb.objects.CombObj.TFBS_from_bed) for precalculated binding sites or [TFBS_from_motifs()](https://tf-comb.readthedocs.io/en/latest/API/tfcomb.objects.html#tfcomb.objects.CombObj.TFBS_from_motifs) for motif scan within regions)?

If the Problem persists please provide an example input data of yours (for example the first 10 lines)  to help me narrow down the problem and I'm happy to assist you further.

Kind Regards
Vanessa

**Comment by dpwlek1997** [comment]
Hi Vanessa, 

Thanks for your kind replies. 

I have recalculated binding sites as bed files which look like below and the name is the name of the gene. 
And since my data is TFBS, i used TFBS_from_bed() function. 

scaffold_12	80264392	80265392	Lyl1	.	.
scaffold_18	1346760	1347760	Ell2	.	.
scaffold_3	98494234	98495234	Nod1	.	.
scaffold_10	25100259	25101259	Rbms1	.	.
scaffold_15	62851021	62852021	Zfp12	.	.
scaffold_6	97132151	97133151	Fam83a	.	.
scaffold_1	204066005	204067005	Traf3	.	.
scaffold_12	96625865	96626865	Tk2	.	.
scaffold_15	40473592	40474592	Kazn	.	.
scaffold_4	112000683	112001683	Tomm70a	.	.
.... 


`# set up an empty CombOb
C = CombObj()

# read TFBS from bed files
C.TFBS_from_bed("") # Tfcomb_test
C.TFBS[:10]

# perform market basket analysis = co-occurrence analysis of the sites.
C.count_within()
C.market_basket()`

C.TFBS[:10]
[scaffold_1	1156137	1157137	Napb	.	.,
 scaffold_1	3905323	3906323	Crnkl1	.	.,
 scaffold_1	3974753	3975753	Naa20	.	.,
 scaffold_1	5316512	5317512	Dzank1	.	.,
 scaffold_1	5467569	5468569	Kat14	.	.,
 scaffold_1	5570573	5571573	Snx5	.	.,
 scaffold_1	5826782	5827782	Rrbp1	.	.,
 scaffold_1	5916495	5917495	Dstn	.	.,
 scaffold_1	6679085	6680085	Snrpb2	.	.,
 scaffold_1	8929196	8930196	Flrt3	.	.]
 
 but still, the rules have zero TF-TF binding. Hope to get more responses from you soon!

**Comment by dpwlek1997** [comment]
@vheger  Thank you so much for your replies!! It was truly helpful to understand your tool much better.
I just ran the code with a 1bp window with default parameters but still got 0 TF pairs... 

```{python}
# INFO: Read 2076 sites (2053 unique names)

# C.TFBS[:10]
scaffold_12	80264893	80264894	Lyl1	.	.
scaffold_18	1347261	1347262	Ell2	.	.
scaffold_3	98494735	98494736	Nod1	.	.
scaffold_10	25100760	25100761	Rbms1	.	.
scaffold_15	62851522	62851523	Zfp12	.	.
scaffold_6	97132652	97132653	Fam83a	.	.
scaffold_1	204066506	204066507	Traf3	.	.
scaffold_12	96626366	96626367	Tk2	.	.
scaffold_15	40474093	40474094	Kazn	.	.
scaffold_4	112001184	112001185	Tomm70a	.	.
```
For further background, this is the data from CutandRun techniques and i applied Diffbind to extract the TF binding sites.

**Comment by dpwlek1997** [comment]
I have also tried with homer-output (TF-motif) -> bed format and TFcomb. 
But here again, got no single TF pair. Do we have to include the background also?  such as not significant TF-motifs. 

```python
INFO: Read 2437 sites (16 unique names)
[scaffold_1	3340643	3351777	LRF(Zf)	.	.,
 scaffold_1	3969056	3979317	Maz(Zf)	.	.,
 scaffold_1	3969056	3979317	Maz(Zf)	.	.,
 scaffold_1	5566633	5579848	BMYB(HTH)	.	.,
 scaffold_1	5913095	5921395	Maz(Zf)	.	.,
 scaffold_1	6816340	6829667	Maz(Zf)	.	.,
 scaffold_1	8925652	8937575	CHR(?)	.	.,
 scaffold_1	12180347	12191855	Nkx6.1(Homeobox)	.	.,
 scaffold_1	12393609	12401041	Nkx2.1(Homeobox)	.	.,
 scaffold_1	18390614	18401946	TEAD2(TEA)	.	.]
```

**Comment by vheger** [comment]
Yes you are right, TF-COMB is designed for but not limited to TFs. And no you don't need to include a background, only the candidates you are interested in. 

What is the exact biological question you want to answer with your data?  I get that you have histone markers, but are you interested if the marker co-occur, or if they co-occur with specific TFs as in our paper, or a complete different question? This would help a lot to help you archive your goal, could you explain a bit more ?

2187 is a rather small number of sites and the loci seem far away from each other, is it possible you don't have two binding sites within a range of 100 bp (approximate average of one regulatory unit)? Would it suit your question to increase the allowed distance (e.g. with  `.count_within(max_dist = 10000)`)? If an increased range is suitable depends on the question you want to answer, technically with a large enough range you should start to get results, as you can see in the attached screenshot (I used only the ten TFBS from your second post and such a high range to illustrate the principle, these settings won't make sense in a real analysis):
![image](https://github.com/loosolab/TF-COMB/assets/85618292/8ba4e820-077c-41a3-9f8a-b5dab89e3691)



For input data: The data we used is public available. You can use the [example data](https://github.com/loosolab/TF-COMB/tree/main/data) to play around (this is a small subset of the data we used) or download full data from ENCODE, a list of all accession numbers is attached to the paper in [supplementary  file 2](https://ars.els-cdn.com/content/image/1-s2.0-S2001037022003051-mmc2.pdf) for the more complex questions we addressed in the paper 🙂  .

**Comment by dpwlek1997** [comment]
First, very appreciated for your detailed and sincere replies. 

Here are some more background and research questions or things I would like to get from the TFcomb Package. Please let me know if the below description is not clear enough. 

< experiment > 
Data
- H3K4me3 on a control sample
- H3K4me3 on a case sample
- H3K27me3 on a control sample
- H3K27me3 on a case sample 

< Research Questions >
  1. Differential co-occurrence of binding sites between conditions (case vs control) for each histone marker.
  2. Co-occurring TF pairs for each experiment (different histone markers in each condition) and also differential co-occurrence of TF pairs between conditions. 

I am more interested in the second question focused on TFs and I used the outputs of homer findMotifsGenome.pl.  And when I increased the max distance to 50000, I was able to get some TF pairs (hopefully enhancers) even if cosine similarity values are not high.  For the first question, I am a bit off the output with the same size as the window. To give you more background, the average peak width is pretty wide with an average of 6000-7000, and afraid this might cause a problem when I narrowed it down to 1 bp. Maybe, should I keep the window size and use max_overlap=1? What's your thought? I am also thinking about combining binding sites from two different markers withing the same sample. (for example, Differential binding sites from H3K4me3 in control and H3K27me3 in control).

And, lastly, could you explain a bit about how you calculated the cosine similarity? 
I am guessing that the cosine value might be computed as TF1_TF2_count/square_root(TF1_count x TF2_count). But from one of my examples, I got a cosine of 1.414214 which resulted from TF1_TF2count=2, TF1_count=2, TF2_count=1 and wasn't sure how TF2_count=2 having less count than pairs. (happened when i used max_dist = 50000)

```python
TF1 | TF2 | TF1_TF2_count | TF1_count | TF2_count | cosine | zscore
-- | -- | -- | -- | -- | -- | --
Irgm1 | Trim41 | 2 | 2 | 1 | 1.414214 | inf
1110038F14Rik | Rpl8 | 1 | 1 | 1 | 1.000000 | inf
```
Thanks a lot for your consideration.

**Comment by vheger** [comment]
To the questions on the research questions, I understand that we can split it into co-occurring histone marks and co-occurring TFs (within different histone marks). 

**For the co-occurring histone marks**: Yes as you mention, it would be beneficial to include all the different marks within the same analysis. If you rename the 4th column of your .bed-files to `H3K4me3_control`, `H3K4me3_case`,  `H3K27me3_control` and `H3K27me3_case`, you can add the sites to the same object using multiple `C.TFBS_from_bed("")` calls (these will append to the list of sites). That way you will have all four groups within the same analysis to calculate the co-occurrence between H3K4me3_control and H3K27me3_control. For this, yes I think it would be good to set max_overlap=1 and adjust the max_dist as needed.

**For the co-occurring TFs**: I am not sure how “_findMotifsGenome.pl_” works. It seems like it is finding very few binding sites and that this is the reason why you are not finding any co-occurring TFs. Do you have a list of motifs you would like to investigate? Then you might also use the TF-COMB function “_TFBS_from_motifs_” as described here: https://tf-comb.readthedocs.io/en/latest/examples/TFBS_from_motifs.html . You could for example run this multiple times, one for each histone mark, and then compare the TF-TF co-occurrences as seen here: https://tf-comb.readthedocs.io/en/latest/examples/Differential_analysis.html 


**Regarding cosine calculation**. We calculate [cosine](https://tf-comb.readthedocs.io/en/latest/methods/mba.html#cosine) as `support(TFA,TFB)/square_root(support(TFA) x support(TFB))`  with [support](https://tf-comb.readthedocs.io/en/latest/methods/mba.html#support) defined as`support(TFA) = TFA_count/number_baskets`. 


For the second part of your question, this happens, if you encounter one (or both) partner(s) multiple times within a window. 
Imagine we have the following situation in our window we count in: `<start>...Irgm1(A)…..Irgm1(B)…Trim41…<end>` (the dots indicate bps in between and I added a (A) and (B) to Irgm1 to distinguish in my explanation, but both are Irgm1). In this scenario we would count Trim41 once and Irgm1 twice (A and B), right? So we get `count_Trim41=1` and `count_Irgm1=2`. We also observe the pairing `Irgm1(A)-Trim41` and `Irgm1(B)-Trim41` within this window, resulting in `count_Irgm1-Trim41=2`, which is the default behaviour. To count the pair only once, you can set [C.count_within(binarize=True)](https://tf-comb.readthedocs.io/en/latest/API/tfcomb.objects.html#tfcomb.objects.CombObj.count_within).

**Comment by ybaeus** [comment]
Here I came up with much better outputs! 

<img width="508" alt="image" src="https://github.com/loosolab/TF-COMB/assets/47170687/b5e5b553-120d-4367-a82a-ea0596008e38">

Here are some examples of how I approached computing cosine based on your explanation. **since they are not scaled within -1 + 1, I am guessing that we can have some results over 1**. Am I right?

#### Data 
TF1 | TF2 | TF1_TF2_count | TF1_count | TF2_count | cosine | zscore
-- | -- | -- | -- | -- | -- | --
Dlx3(Homeobox) | Nkx6.1(Homeobox) | 859 | 301 | 864 | 1.684430 | 46.514646

<CombObj: 7656 TFBS (16 unique names) | Market basket analysis: 256 rules>

#### Cosine computation
support(TFA,TFB) = 859 / 256 
square_root(support(TFA) x support(TFB)) = square_root(301/256 x 864/256)

support(TFA,TFB)/square_root(support(TFA) x support(TFB)) ~ 1.6844 YEAH~!! Also had tested with other pairs and was fit. 

And also did test for differential analysis which resulted 
![image](https://github.com/loosolab/TF-COMB/assets/47170687/c39e01bf-a6b8-4b69-b8c1-40599b589046)
They had only 49 shared pairs so not a lot of dots but please feel free to give any feedbacks on it if possible
